Documentation/bpf.rst

```diff
@@ -1181,6 +1181,64 @@ The generated LLVM IR can also be dumped in human readable format through:
 
     $ clang -O2 -Wall -emit-llvm -S -c xdp-example.c -o -
 
+In addition, For conversion debugging information to BTF (BPF Type Format),
+more recent ``llc`` and ``pahole`` are needed and ``llc`` and ``pahole``
+can be probed for BTF support:
+
+::
+
+    $ llc -march=bpf -mattr=help |& grep dwarfris
+      dwarfris - Disable MCAsmInfo DwarfUsesRelocationsAcrossSections.
+      dwarfris - Disable MCAsmInfo DwarfUsesRelocationsAcrossSections.
+      dwarfris - Disable MCAsmInfo DwarfUsesRelocationsAcrossSections.
+    $ pahole --help | greo BTF
+    -J, --btf_encode           Encode as BTF
+
+Recent pahole version can be gotten from https://git.kernel.org/pub/scm/devel/pahole/pahole.git
+and note that the ``llvm-objcopy`` tool is required for ``pahole``, so probe it, too.
+
+``-mattr=dwarfris`` should be used with llc and enables ``pahole`` to dump structures correctly.
+(For ``pahole``, dwarf cross-section relation should be disabled through ``-mattr=dwarfris``,
+because BPF backend is not yet implemented in elfutils backend)
+
+Like the above two steps, generating a binary LLVM IR intermediate file
+``xdp-example.bc`` with debugging informantion by adding ``-g``, which
+can later on be passed to llc with ``-mattr=dwarfris``:
+
+::
+
+    $ clang -O2 -g -Wall -target bpf -emit-llvm -c xdp-example.c -o xdp-example.bc
+    $ llc xdp-example.bc -march=bpf -mattr=dwarfris -filetype=obj -o xdp-example.o
+
+And then ``pahole`` can properly dump structures of the BPF program:
+
+::
+
+    $ pahole xdp-example.o
+    struct xdp_md {
+            __u32                      data;                 /*     0     4 */
+            __u32                      data_end;             /*     4     4 */
+            __u32                      data_meta;            /*     8     4 */
+
+            /* size: 12, cachelines: 1, members: 3 */
+            /* last cacheline: 12 bytes */
+    };
+
+Through the option ``-J``, ``pahole`` can convert dwarf format to BTF:
+
+::
+
+    $ pahole -J xdp-example.o
+
+The ``.BTF`` section can be dumped through readelf:
+
+::
+
+    $ readelf -a xdp-example.o
+    [...]
+      [18] .BTF              PROGBITS         0000000000000000  00000671
+    [...]
+
```

Comment:
Additionally, it would be good to add an example on how to use BTF in combination with BPF maps with the `BPF_ANNOTATE_KV_PAIR()` macro, and how this data can be extracted via bpftool. I've put this into the commit description here: https://git.kernel.org/pub/scm/network/iproute2/iproute2-next.git/commit/?id=f823f36012fb5ab4ddfca6ed4ff56188730f281e

I think this could be a commit on top of this one and probably should go into iproute2 section, but it would be really useful to demonstrate here in this context.

In any case, thanks a lot for the patch!

Reply:
+1

Having an example on how BTF getting consumed (e.g. by "bpftool map dump id 123") will be useful.

Reply:
OK, I also think it'd be great and useful to add the above examples you said,
but I need some time to test such examples from various angles. After the tests is done, will send v2 :smile: 
Thanks for detailed review , kafai , borkmann !!

Reply:
@borkmann , I have a question. I tested the example on my machine like below:

Where the below ouput of `bpftool map dump id 26` actually is printed ? which source code file ?
I thought "tools/bpf/bpftool" have not some code printing the output, no ?

(I cannot figure out the internal logic how to print the output based on btf info showing member variables of `struct ctl_value`)
```c
taeung@linux:~$ sudo ip -force link set dev lo xdp obj test_xdp_noinline.o sec xdp-test
taeung@linux:~$ ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 xdpgeneric/id:11 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
[...]
taeung@linux:~/bpf/bpftool$ sudo ./bpftool prog show id 11
11: xdp  tag c154b9fcbc44bd95  gpl
	loaded_at 2018-07-31T02:03:23+0900  uid 0
	xlated 8144B  not jited  memlock 12288B  map_ids 21,25,26,22,24,23
taeung@linux:~/bpf/bpftool$ sudo ./bpftool map dump id 26
[{
        "key": 0,
        "value": {
            "": {
                "value": 0,
                "ifindex": 0,
                "mac": []
            }
        }
    },{
        "key": 1,
        "value": {
            "": {
                "value": 0,
                "ifindex": 0,
                "mac": []
            }
        }
    },{
[...]
```

Reply:
@Taeung  FYI. The original patch which added this functionality is in bpf/net tree.
```
commit 38d5d3b3d5dbc0e0bb51fa6f7559d0d5a27916f6
Author: Martin KaFai Lau <kafai@fb.com>
Date:   Tue Jul 24 08:40:22 2018 -0700

    bpf: Introduce BPF_ANNOTATE_KV_PAIR
    
    This patch introduces BPF_ANNOTATE_KV_PAIR to signal the
    bpf loader about the btf key_type and value_type of a bpf map.
    Please refer to the changes in test_btf_haskv.c for its usage.
    Both iproute2 and libbpf loader will then have the same
    convention to find out the map's btf_key_type_id and
    btf_value_type_id from a map's name.
    
    Fixes: 8a138aed4a80 ("bpf: btf: Add BTF support to libbpf")
    Suggested-by: Daniel Borkmann <daniel@iogearbox.net>
    Signed-off-by: Martin KaFai Lau <kafai@fb.com>
    Acked-by: Yonghong Song <yhs@fb.com>
    Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>
```

Reply:
Thanks, @borkmann @yonghong-song !!
Now I got it the internal logic of BTF dumping.
The reason I confused was because the gap regarding this example between `bpf-next` and `bpf` branch..

(bpftool `do_dump_btf()`(for this example) is implemented on only `bpf-next` tree,
but there is the `BPF_ANNOTATE_KV_PAIR` of bpf_helpers.h  on `bpf` tree) :cry: 
Anyway, will send v2 soon !!